File: togo/internal/packages/version_range.go
package packages

import (
	"math/big"
	"regexp"
	"strings"
)

// Maven version-range support. Port of src/packages/versionRange.ts.
//
// cappu otherwise treats a declared version as an exact coordinate; a Maven
// range (`[1.0,2.0)`) - in cappu.json or a real-world transitive POM - must
// first be resolved to a concrete published version. This file provides the
// ordering (a subset of Maven's ComparableVersion), the range parser,
// membership, and the "highest published match" selector.
//
// Scope: bracket/paren ranges, comma-joined sets, and the RELEASE/LATEST
// tokens. A bare version keeps its exact-pin meaning (ParseVersionSpec returns
// ok=false for it) and is NOT reinterpreted as a Maven soft requirement.

// Known qualifier ranks (lower sorts earlier). "" is the release itself and
// outranks every pre-release qualifier; "sp" is >= release.
// ponytail: Maven's full alias table is reproduced only for the common aliases
// below; an unknown qualifier sorts after the release and lexically among its
// peers - upgrade the table if a real dependency's ordering needs it.
var qualifierRank = map[string]int{
	"alpha":     -6,
	"a":         -6,
	"beta":      -5,
	"b":         -5,
	"milestone": -4,
	"m":         -4,
	"rc":        -3,
	"cr":        -3,
	"snapshot":  -2,
	"":          0, // the release
	"ga":        0,
	"final":     0,
	"release":   0,
	"sp":        1,
}

type segment struct {
	num  *big.Int // nil for a qualifier segment
	qual string
}

var (
	splitRe = regexp.MustCompile(`[.\-_+]`)
	pieceRe = regexp.MustCompile(`\d+|[a-z]+`)
	digitRe = regexp.MustCompile(`^\d+$`)
)

// segmentsOf splits into numeric and qualifier segments. Maven separates on `.`
// and `-`, and also at a digit/letter transition ("1alpha" -> "1", "alpha").
func segmentsOf(version string) []segment {
	var segments []segment
	for _, raw := range splitRe.Split(strings.ToLower(version), -1) {
		if raw == "" {
			continue
		}
		for _, piece := range pieceRe.FindAllString(raw, -1) {
			if digitRe.MatchString(piece) {
				n := new(big.Int)
				n.SetString(piece, 10)
				segments = append(segments, segment{num: n})
			} else {
				segments = append(segments, segment{qual: piece})
			}
		}
	}
	return segments
}

func sign(n int) int {
	switch {
	case n < 0:
		return -1
	case n > 0:
		return 1
	default:
		return 0
	}
}

// compareQualifier compares two qualifiers by known rank, then lexically; a
// known qualifier always sorts before an unknown one.
func compareQualifier(aq, bq string) int {
	ar, aok := qualifierRank[aq]
	br, bok := qualifierRank[bq]
	switch {
	case aok && bok:
		return sign(ar - br)
	case aok:
		return -1
	case bok:
		return 1
	}
	return strings.Compare(aq, bq)
}

// compareToMissing compares a present segment against a missing one: a number
// compares against 0 (so 1.0 == 1.0.0), a qualifier against the release.
func compareToMissing(s segment) int {
	if s.num != nil {
		return s.num.Sign()
	}
	return compareQualifier(s.qual, "")
}

func compareSegment(a, b *segment) int {
	switch {
	case a == nil && b == nil:
		return 0
	case a == nil:
		return -compareToMissing(*b)
	case b == nil:
		return compareToMissing(*a)
	}
	if a.num != nil && b.num != nil {
		return a.num.Cmp(b.num)
	}
	// A number always outranks a qualifier (Maven: 1.1 > 1.1-alpha).
	if a.num != nil {
		return 1
	}
	if b.num != nil {
		return -1
	}
	return compareQualifier(a.qual, b.qual)
}

// CompareVersions gives Maven-style ordering: negative if a<b, positive if a>b.
func CompareVersions(a, b string) int {
	as := segmentsOf(a)
	bs := segmentsOf(b)
	n := len(as)
	if len(bs) > n {
		n = len(bs)
	}
	at := func(s []segment, i int) *segment {
		if i < len(s) {
			return &s[i]
		}
		return nil
	}
	for i := 0; i < n; i++ {
		if c := compareSegment(at(as, i), at(bs, i)); c != 0 {
			return c
		}
	}
	return 0
}

type restriction struct {
	lower          string
	lowerInclusive bool
	upper          string
	upperInclusive bool
	hasLower       bool
	hasUpper       bool
}

// VersionSpec is a parsed Maven version range or newest-wins token.
type VersionSpec struct {
	// Newest is set for RELEASE/LATEST: pick the highest published version.
	Newest bool
	// Restrictions are OR-joined; a version satisfies the spec if it satisfies any.
	restrictions []restriction
}

func parseRestriction(text string) (restriction, bool) {
	lowerInclusive := strings.HasPrefix(text, "[")
	upperInclusive := strings.HasSuffix(text, "]")
	open := strings.HasPrefix(text, "[") || strings.HasPrefix(text, "(")
	closed := strings.HasSuffix(text, "]") || strings.HasSuffix(text, ")")
	if !open || !closed {
		return restriction{}, false
	}
	inner := text[1 : len(text)-1]
	if !strings.Contains(inner, ",") {
		// [1.5] - a single hard version
		if !lowerInclusive || !upperInclusive || inner == "" {
			return restriction{}, false
		}
		return restriction{
			lower: inner, lowerInclusive: true, hasLower: true,
			upper: inner, upperInclusive: true, hasUpper: true,
		}, true
	}
	comma := strings.Index(inner, ",")
	lower := strings.TrimSpace(inner[:comma])
	upper := strings.TrimSpace(inner[comma+1:])
	return restriction{
		lower: lower, lowerInclusive: lowerInclusive, hasLower: lower != "",
		upper: upper, upperInclusive: upperInclusive, hasUpper: upper != "",
	}, true
}

// ParseVersionSpec parses a Maven version spec. ok is false when spec is a plain
// exact version (the caller then treats it as an exact coordinate, as before).
func ParseVersionSpec(spec string) (VersionSpec, bool) {
	trimmed := strings.TrimSpace(spec)
	if trimmed == "RELEASE" || trimmed == "LATEST" {
		return VersionSpec{Newest: true}, true
	}
	if !strings.HasPrefix(trimmed, "[") && !strings.HasPrefix(trimmed, "(") {
		return VersionSpec{}, false
	}
	var restrictions []restriction
	depth := 0
	start := 0
	for i := 0; i <= len(trimmed); i++ {
		var ch byte
		if i < len(trimmed) {
			ch = trimmed[i]
		}
		switch ch {
		case '[', '(':
			depth++
		case ']', ')':
			depth--
		}
		if i == len(trimmed) || (ch == ',' && depth == 0) {
			part := strings.TrimSpace(trimmed[start:i])
			if part != "" {
				r, ok := parseRestriction(part)
				if !ok {
					return VersionSpec{}, false // malformed: fall back to exact
				}
				restrictions = append(restrictions, r)
			}
			start = i + 1
		}
	}
	if len(restrictions) == 0 {
		return VersionSpec{}, false
	}
	return VersionSpec{restrictions: restrictions}, true
}

func withinRestriction(r restriction, version string) bool {
	if r.hasLower {
		c := CompareVersions(version, r.lower)
		if c < 0 || (c == 0 && !r.lowerInclusive) {
			return false
		}
	}
	if r.hasUpper {
		c := CompareVersions(version, r.upper)
		if c > 0 || (c == 0 && !r.upperInclusive) {
			return false
		}
	}
	return true
}

// Satisfies reports whether version satisfies spec (RELEASE/LATEST accept any).
func Satisfies(spec VersionSpec, version string) bool {
	if spec.Newest {
		return true
	}
	for _, r := range spec.restrictions {
		if withinRestriction(r, version) {
			return true
		}
	}
	return false
}

// SelectVersion returns the highest published version satisfying spec (Maven
// picks the highest in range; RELEASE/LATEST pick the newest), or "" if none.
func SelectVersion(spec VersionSpec, published []string) string {
	best := ""
	for _, version := range published {
		if !Satisfies(spec, version) {
			continue
		}
		if best == "" || CompareVersions(version, best) > 0 {
			best = version
		}
	}
	return best
}

File: src/packages/versionRange.ts
// Maven version-range support. cappu otherwise treats a declared version as an
// exact coordinate; a Maven range (`[1.0,2.0)`) - in cappu.json or a real-world
// transitive POM - must first be resolved to a concrete published version.
// This module provides the ordering (a subset of Maven's ComparableVersion),
// the range parser, membership, and the "highest published match" selector.
//
// Scope: bracket/paren ranges, comma-joined sets, and the RELEASE/LATEST
// tokens. A bare version keeps its exact-pin meaning (parseVersionSpec returns
// undefined for it) and is NOT reinterpreted as a Maven soft requirement.

// ---- ordering (Maven ComparableVersion, subset) -----------------------------

// Known qualifier ranks (lower sorts earlier). "" is the release itself and
// outranks every pre-release qualifier; "sp"/"ga"/"final" are >= release.
// ponytail: Maven's full alias table (a=alpha, b=beta, m=milestone, cr=rc, ...)
// is reproduced only for the common aliases below; an unknown qualifier sorts
// after the release and lexically among its peers - upgrade the table if a real
// dependency's ordering needs it.
const QUALIFIER_RANK = new Map<string, number>([
  ["alpha", -6],
  ["a", -6],
  ["beta", -5],
  ["b", -5],
  ["milestone", -4],
  ["m", -4],
  ["rc", -3],
  ["cr", -3],
  ["snapshot", -2],
  ["", 0], // the release
  ["ga", 0],
  ["final", 0],
  ["release", 0],
  ["sp", 1],
]);

type Segment = { readonly num?: bigint; readonly qual?: string };

// Split into numeric and qualifier segments. Maven separates on `.` and `-`,
// and also at a digit/letter transition ("1alpha" -> "1", "alpha"). A numeric
// segment is compared numerically; a qualifier segment by rank then lexically.
function segmentsOf(version: string): Segment[] {
  const segments: Segment[] = [];
  for (const raw of version.toLowerCase().split(/[.\-_+]/)) {
    if (raw === "") continue;
    // break at each digit<->letter boundary
    for (const piece of raw.match(/\d+|[a-z]+/g) ?? []) {
      if (/^\d+$/.test(piece)) segments.push({ num: BigInt(piece) });
      else segments.push({ qual: piece });
    }
  }
  return segments;
}

// Compare two qualifiers by known rank, then lexically; a known (pre-release or
// release) qualifier always sorts before an unknown one (Maven ranks unknown
// qualifiers after the release).
function compareQualifier(aq: string, bq: string): number {
  const ar = QUALIFIER_RANK.get(aq);
  const br = QUALIFIER_RANK.get(bq);
  if (ar !== undefined && br !== undefined) return Math.sign(ar - br);
  if (ar !== undefined) return -1;
  if (br !== undefined) return 1;
  return aq < bq ? -1 : aq > bq ? 1 : 0;
}

// A present segment against a missing one (the shorter version padded): a
// number compares against 0 (so 1.0 == 1.0.0), a qualifier against the release
// (so 1.0-alpha < 1.0, but 1.0-sp > 1.0).
function compareToMissing(seg: Segment): number {
  if (seg.num !== undefined) return seg.num > 0n ? 1 : seg.num < 0n ? -1 : 0;
  return compareQualifier(seg.qual!, "");
}

function compareSegment(a: Segment | undefined, b: Segment | undefined): number {
  if (a === undefined) return b === undefined ? 0 : -compareToMissing(b);
  if (b === undefined) return compareToMissing(a);
  if (a.num !== undefined && b.num !== undefined) return a.num < b.num ? -1 : a.num > b.num ? 1 : 0;
  // A number always outranks a qualifier (Maven: 1.1 > 1.1-alpha).
  if (a.num !== undefined) return 1;
  if (b.num !== undefined) return -1;
  return compareQualifier(a.qual!, b.qual!);
}

/** Maven-style version ordering: negative if a<b, positive if a>b, 0 if equal. */
export function compareVersions(a: string, b: string): number {
  const as = segmentsOf(a);
  const bs = segmentsOf(b);
  const len = Math.max(as.length, bs.length);
  for (let i = 0; i < len; i++) {
    const c = compareSegment(as[i], bs[i]);
    if (c !== 0) return c;
  }
  return 0;
}

// ---- ranges -----------------------------------------------------------------

interface Restriction {
  readonly lower?: string;
  readonly lowerInclusive: boolean;
  readonly upper?: string;
  readonly upperInclusive: boolean;
}

export interface VersionSpec {
  /** Any-newest tokens (RELEASE/LATEST): pick the highest published version. */
  readonly newest?: boolean;
  /** OR-joined restrictions; a version satisfies the spec if it satisfies any. */
  readonly restrictions: readonly Restriction[];
}

function parseRestriction(text: string): Restriction | undefined {
  const lowerInclusive = text.startsWith("[");
  const upperInclusive = text.endsWith("]");
  const open = text.startsWith("[") || text.startsWith("(");
  const close = text.endsWith("]") || text.endsWith(")");
  if (!open || !close) return undefined;
  const inner = text.slice(1, -1);
  if (!inner.includes(",")) {
    // [1.5] - a single hard version (both bounds inclusive on that version)
    if (!lowerInclusive || !upperInclusive || inner === "") return undefined;
    return { lower: inner, lowerInclusive: true, upper: inner, upperInclusive: true };
  }
  const comma = inner.indexOf(",");
  const lower = inner.slice(0, comma).trim();
  const upper = inner.slice(comma + 1).trim();
  return {
    ...(lower !== "" ? { lower } : {}),
    lowerInclusive,
    ...(upper !== "" ? { upper } : {}),
    upperInclusive,
  };
}

/**
 * Parse a Maven version spec, or undefined when `spec` is a plain exact version
 * (the caller then treats it as an exact coordinate, as before). RELEASE/LATEST
 * become a newest-wins spec.
 */
export function parseVersionSpec(spec: string): VersionSpec | undefined {
  const trimmed = spec.trim();
  if (trimmed === "RELEASE" || trimmed === "LATEST") return { newest: true, restrictions: [] };
  if (!trimmed.startsWith("[") && !trimmed.startsWith("(")) return undefined;
  // Split comma-joined sets at the commas *between* restrictions (those that
  // follow a `]` or `)`), leaving the intra-restriction commas intact.
  const restrictions: Restriction[] = [];
  let depth = 0;
  let start = 0;
  for (let i = 0; i <= trimmed.length; i++) {
    const ch = trimmed[i];
    if (ch === "[" || ch === "(") depth++;
    else if (ch === "]" || ch === ")") depth--;
    if (i === trimmed.length || (ch === "," && depth === 0)) {
      const part = trimmed.slice(start, i).trim();
      if (part !== "") {
        const restriction = parseRestriction(part);
        if (!restriction) return undefined; // malformed: fall back to exact
        restrictions.push(restriction);
      }
      start = i + 1;
    }
  }
  return restrictions.length > 0 ? { restrictions } : undefined;
}

function withinRestriction(r: Restriction, version: string): boolean {
  if (r.lower !== undefined) {
    const c = compareVersions(version, r.lower);
    if (c < 0 || (c === 0 && !r.lowerInclusive)) return false;
  }
  if (r.upper !== undefined) {
    const c = compareVersions(version, r.upper);
    if (c > 0 || (c === 0 && !r.upperInclusive)) return false;
  }
  return true;
}

/** Whether `version` satisfies `spec` (RELEASE/LATEST are satisfied by any). */
export function satisfies(spec: VersionSpec, version: string): boolean {
  if (spec.newest) return true;
  return spec.restrictions.some(r => withinRestriction(r, version));
}

/**
 * The highest published version satisfying `spec` (Maven picks the highest in
 * range; RELEASE/LATEST pick the newest overall), or undefined when none match.
 */
export function selectVersion(spec: VersionSpec, published: readonly string[]): string | undefined {
  let best: string | undefined;
  for (const version of published) {
    if (!satisfies(spec, version)) continue;
    if (best === undefined || compareVersions(version, best) > 0) best = version;
  }
  return best;
}
